File: ios/react-native-bluetooth-cross-platform/TransportHandler.swift
import Foundation
import Underdark

open class TransportHandler: RCTEventEmitter, UDTransportDelegate {
  
  // delimiterse
  private var type: User.PeerType = User.PeerType.OFFLINE
  private var transportConfigured: Bool = false
  private var transport: UDTransport? = nil
  private var advertiseTimer: Timer? = nil
  private var nodeId: Int64 = 0
  internal var links = [UDLink]()
  internal var nearbyUsers = [User]()

  // ratkinson - setting up a dispatch queue to handle "links" and "nearbyUsers"
  // logic syncronously in eventHandlers
  let serialQueue = DispatchQueue(label: "serialQueue")
  
  // MARK: START TRANSPORT
  public func initTransport(_ kind: String, inType: User.PeerType) {
    if !self.transportConfigured {
      let appId: Int32 = 234235
      let queue = DispatchQueue.global(qos: DispatchQoS.QoSClass.background)
      var buf : Int64 = 0;
      while buf == 0 {
        arc4random_buf(&buf, MemoryLayout.size(ofValue: buf))
      }
      if buf < 0 {
        buf = -buf
      }
      nodeId = buf
      var transportKinds = [AnyObject]()
      if kind == "WIFI" {
        transportKinds.append(UDTransportKind.wifi.rawValue as AnyObject)
      }
      if kind == "BT" {
        transportKinds.append(UDTransportKind.bluetooth.rawValue as AnyObject)
      }
      if kind == "WIFI-BT" {
        transportKinds.append(UDTransportKind.bluetooth.rawValue as AnyObject)
        transportKinds.append(UDTransportKind.wifi.rawValue as AnyObject)
      }
      transport = UDUnderdark.configureTransport(withAppId: appId, nodeId: nodeId, queue: queue, kinds: transportKinds)
      transport?.delegate = self
      self.transportConfigured = true
    }
    transport?.start()
  }

  // MARK: stop transport
  open func stopTransport() {
    transport?.stop()
  }
  
  // MARK: TRansport delegate
  public func transport(_ transport: UDTransport, linkConnected link: UDLink) {
    serialQueue.sync {
      links.append(link)
    }
  }
  
  public func transport(_ transport: UDTransport, linkDisconnected link: UDLink) {
    serialQueue.sync {
      let linksCount = links.count
      var i = (linksCount - 1);

      while i >= 0 {
        if link.nodeId == links[i].nodeId {
          links.remove(at: i)
        }
        i -= 1
      }

      let nearbyUserCount = nearbyUsers.count
      i = (nearbyUserCount - 1)

      while i >= 0 {
        if link.nodeId == nearbyUsers[i].link.nodeId {
          self.sendEvent(withName: "lostUser", body:  nearbyUsers[i].getJSUser("lost peer"))
          nearbyUsers.remove(at: i)
        }
        i -= 1
      }
    }
  }

  public func transport(_ transport: UDTransport, link: UDLink, didReceiveFrame frameData: Data) {
    // handle this in network communicatorwith proper encoder and decoder functionality
  }
  
  override open func supportedEvents() -> [String]! {
    return ["lostUser","detectedUser", "messageReceived", "connectedToUser", "receivedInvitation"]
  }
  
}

File: android/src/main/java/com/rctunderdark/TransportHandler.java
package com.rctunderdark;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import java.util.EnumSet;
import java.util.Random;
import java.util.Vector;

import io.underdark.Underdark;
import io.underdark.transport.Link;
import io.underdark.transport.Transport;
import io.underdark.transport.TransportKind;
import io.underdark.transport.TransportListener;

public class TransportHandler implements TransportListener {
    // MARK: private variables
    private boolean transportConfigured = false;
    private Transport transport;
    protected Vector<Link> links;
    protected Vector<User> nearbyUsers;
    private TransportListener listener;
    protected ReactContext context;
    private long nodeId = 0;

    // MARK: ReactContextBaseJavaModule
    public TransportHandler(ReactApplicationContext reactContext) {
        this.context = reactContext;
        this.listener = this;
    }
    public void initTransport(String kind, User.PeerType inType) {
        if(transportConfigured){
            transport.start();
            return;
        }
        this.listener = this;
        links = new Vector<Link>();
        nearbyUsers = new Vector<User>();
        while(nodeId == 0) {
            nodeId = new Random().nextLong();
        }
        EnumSet<TransportKind> kinds = null;
        if(nodeId < 0) {
            nodeId = -nodeId;
        }

        if("WIFI".equals(kind)) {
            kinds = EnumSet.of(TransportKind.WIFI);

        } else if("BT".equals(kind)) {
            kinds = EnumSet.of(TransportKind.BLUETOOTH);

        } else {
            kinds = EnumSet.of(TransportKind.WIFI, TransportKind.BLUETOOTH);
        }
        this.transport = Underdark.configureTransport(
                234235,
                nodeId,
                this,
                null,
                context,
                kinds
        );
        this.transportConfigured = true;
        transport.start();
    }
    public void stopTransport() {
        transport.stop();
    }
    //MARK: TransportListener
    @Override
    public void transportNeedsActivity(Transport transport, ActivityCallback callback) {

    }
    @Override
    public void transportLinkConnected(Transport transport, Link link) {
        this.links.add(link);
    }
    @Override
    public void transportLinkDisconnected(Transport transport, Link link) {
        int i = 0;
        while(i < links.size()) {
            if(link.getNodeId() == links.elementAt(i).getNodeId()) {
                this.links.removeElementAt(i);
            } else {
                ++i;
            }
        }
        i = 0;
        while(i < nearbyUsers.size()) {
            if (link.getNodeId() != nearbyUsers.elementAt(i).link.getNodeId()) {
                ++i;
            } else {
                WritableMap map = nearbyUsers.elementAt(i).getJSUser();
                map.putString("message", "lost peer");
                context.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                        .emit("lostUser", map);
                this.nearbyUsers.removeElementAt(i);;
            }
        }
    }
    @Override
    public void transportLinkDidReceiveFrame(Transport transport, Link link, byte[] frameData) {
        if (link.getNodeId() == this.nodeId) {
            return;
        }
        // handle this in network communicatorwith proper encoder and decoder functionality
    }
}
